vicky/src/lib/database/entities/task.rs

```
@@ -0,0 +1,415 @@
+            id: self.id.unwrap_or_else(Uuid::new_v4),
+            display_name: self.display_name.unwrap_or_else(|| "Task".to_string()),
+            features: self.features,
+            status: self.status,
+            locks: self.locks,
+            flake_ref: self.flake_ref,
+        }
+    }
+}
+
+// this was on purpose because these macro-generated entity types
+// mess up the whole namespace and HAVE to be scoped
+pub mod db_impl {
+    use crate::database::entities::lock::Lock;
+    use crate::database::entities::task::{Task, TaskResult, TaskStatus};
+    use crate::database::entities::FlakeRef;
+    use crate::errors::VickyError;
+    use diesel::{
+        insert_into, AsChangeset, ExpressionMethods, Identifiable, Insertable, QueryDsl, Queryable,
+        RunQueryDsl, Selectable,
+    };
+    use std::collections::HashMap;
+    use std::fmt::Display;
+    use uuid::Uuid;
+    // these here are evil >:(
+    use crate::database::schema::locks;
+    use crate::database::schema::tasks;
+    use itertools::Itertools;
+    use rocket_sync_db_pools::database;
+
+    #[derive(Insertable, Queryable, AsChangeset, Debug)]
+    #[diesel(table_name = tasks)]
+    struct DbTask {
+        pub id: Uuid,
+        pub display_name: String,
+        pub status: String,
+        pub features: String,
+        pub flake_ref_uri: String,
+        pub flake_ref_args: String,
+    }
+
+    impl Display for TaskStatus {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            let str = match self {
+                TaskStatus::NEW => "NEW",
+                TaskStatus::RUNNING => "RUNNING",
+                TaskStatus::FINISHED(r) => match r {
+                    TaskResult::SUCCESS => "FINISHED::SUCCESS",
+                    TaskResult::ERROR => "FINISHED::ERROR",
+                },
+            };
+            write!(f, "{}", str)
+        }
+    }
+
+    impl From<&str> for TaskStatus {
+        fn from(str: &str) -> TaskStatus {
```

We might want to use TryFrom here and fail explicitly, if we see a str, which is not associated with a task state and not default to `NEW`

---

Good idea.